### PR TITLE
Added CNN to networks and QFunction class

### PR DESCRIPTION
**Pull request by ahtsan**

### CODE REVIEW

File: garage/tf/q_functions/continuous_mlp_q_function.py [hunk 1]
```diff
@@ -5,10 +5,10 @@
 import garage.tf.core.layers as L
 from garage.tf.core.layers import batch_norm
 from garage.tf.misc import tensor_utils
-from garage.tf.q_functions import QFunction
+from garage.tf.q_functions import NetworkBuilder
```
Comment (CatherineSue) [inline garage/tf/q_functions/continuous_mlp_q_function.py]: Where is NetworkBuilder? Am I missing something here?

File: garage/tf/q_functions/mlp_q_function.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import tensorflow as tf
+
+from garage.misc.overrides import overrides
+from garage.tf.core.networks import mlp
+from garage.tf.q_functions import QFunction
+
+
+class MLPQFunction(QFunction):
```
Comment (CatherineSue) [inline garage/tf/q_functions/mlp_q_function.py]: Can we replace ContinuousMLPQFunction with MLPQFunction?

File: garage/tf/core/networks.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
 """
-Multilayer Perceptrons (MLP) with tensorflow as the only dependency.
+Different networks with tensorflow as the only dependency.
 
-The module contains MLP which serves as the base of all networks.
+The module contains MLP and CNN implementation.
 It aims to replace existing implementation of MLP class
```
Comment (CatherineSue) [inline garage/tf/core/networks.py]: Perhaps delete MLP?

File: garage/tf/q_functions/continuous_mlp_q_function.py (hunk 1 above)
Comment (ahtsan) [reply garage/tf/q_functions/continuous_mlp_q_function.py]: oh I renamed QFunction to NetworkBuilder before
I will change that back

File: garage/tf/core/networks.py (hunk 3 above)
Comment (ahtsan) [reply garage/tf/core/networks.py]: why?

File: garage/tf/q_functions/mlp_q_function.py (hunk 2 above)
Comment (ahtsan) [reply garage/tf/q_functions/mlp_q_function.py]: What does the merge layer do in ContinuousMLPQFunction?

File: garage/tf/core/networks.py (hunk 3 above)
Comment (CatherineSue) [reply garage/tf/core/networks.py]: I meant delete 'MLP' in line 5.

Comment (ahtsan) [reply garage/tf/core/networks.py]: oh yes

File: garage/tf/q_functions/continuous_mlp_q_function.py (hunk 1 above)
Comment (ahtsan) [reply garage/tf/q_functions/continuous_mlp_q_function.py]: Done

File: garage/tf/core/networks.py (hunk 3 above)
Comment (ahtsan) [reply garage/tf/core/networks.py]: Done

File: garage/tf/q_functions/__init__.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 from garage.tf.q_functions.base import QFunction
 from garage.tf.q_functions.continuous_mlp_q_function import (
     ContinuousMLPQFunction)
+from garage.tf.q_functions.mlp_q_function import MLPQFunction
```
[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/q_functions/__init__.py]: yes

File: garage/tf/core/networks.py [hunk 5]
```diff
@@ -100,3 +102,98 @@ def parameter(input_var,
             axis=0, values=[tf.shape(input_var)[:ndim - 1], [length]])
         p_broadcast = broadcast_to(p, shape=broadcast_shape)
         return p_broadcast
+
+
+def cnn(input_var,
```
[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/core/networks.py]: should I create garage.tf.networks? or just put the mlp.py/cnn.py in garage.tf.core

[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/core/networks.py]: if I do that, how about the parameter function? Should I also do parameter.py?

File: garage/tf/core/cnn.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+def cnn(input_var,
+        output_dim,
+        filter_dims,
+        num_filters,
+        stride,
+        name,
+        padding="SAME",
+        max_pooling=False,
+        pool_shape=(2, 2),
```
Comment (ryanjulian) [inline garage/tf/core/cnn.py]: Is this really a sane default (i.e. that's what people usually use) or should we force the user to specify?

File: garage/tf/q_functions/continuous_mlp_q_function.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahtsan) [reply garage/tf/q_functions/continuous_mlp_q_function.py]: That class doesn't exist. I updated.

File: garage/tf/core/mlp.py [hunk 7]
```diff
@@ -1,12 +1,11 @@
 """
-Multilayer Perceptrons (MLP) with tensorflow as the only dependency.
+MLP with tensorflow as the only dependency.
 
-The module contains MLP which serves as the base of all networks.
-It aims to replace existing implementation of MLP class
+The module contains MLP implementation.
+It aims to replace existing implementation of network class
```
Comment (ryanjulian) [inline garage/tf/core/mlp.py]: Editorial comments about the development process aren't usually necessary in documentation.

File: garage/tf/core/parameter.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+
+def parameter(input_var,
+              length,
+              initializer=tf.zeros_initializer(),
+              dtype=tf.float32,
+              trainable=True,
+              name="parameter"):
+    """
+    Paramter function.
```
Comment (ryanjulian) [inline garage/tf/core/parameter.py]: It's not a function. It's either a network or a layer.

File: garage/tf/core/parameter.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    It creates variables that could be
```
Comment (ryanjulian) [inline garage/tf/core/parameter.py]: Weird wrapping here.

File: garage/tf/core/parameter.py (hunk 8 above)
Comment (ryanjulian) [reply garage/tf/core/parameter.py]: Perhaps explain why we would want such a thing

File: garage/tf/q_functions/base.py [hunk 10]
```diff
@@ -1,5 +1,25 @@
+import tensorflow as tf
+
 from garage.tf.core import Parameterized
 
 
 class QFunction(Parameterized):
-    pass
+    def _build_net(self, name, input_var):
```
Comment (ryanjulian) [inline garage/tf/q_functions/base.py]: Making a private function part of a public interface is a strange choice. We will have to revisit this as we define TensorFlow models.

File: garage/tf/q_functions/discrete_mlp_q_function.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import tensorflow as tf
+
+from garage.misc.overrides import overrides
+from garage.tf.core.mlp import mlp
+from garage.tf.q_functions import QFunction
+
+
+class DiscreteMLPQFunction(QFunction):
+    """
+    Discrelte MLP Functoin class.
```
Comment (ryanjulian) [inline garage/tf/q_functions/discrete_mlp_q_function.py]: Function

File: garage/tf/q_functions/discrete_mlp_q_function.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    This class implements a q value network to predict q based on the input
```
Comment (ryanjulian) [inline garage/tf/q_functions/discrete_mlp_q_function.py]: Please be consistent with literature in your comments

Q-value network. Predicts the Q-value based on....

File: tests/garage/tf/core/test_cnn.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            self.cnn = cnn(
+                input_var=self._input_ph,
+                output_dim=self._output_shape,
+                filter_dims=(3, 3, 3),
+                num_filters=(32, 64, 128),
+                stride=1,
+                name="cnn1",
+                hidden_nonlinearity=self.hidden_nonlinearity)
+
+        self.sess.run(tf.global_variables_initializer())
```
Comment (ryanjulian) [inline tests/garage/tf/core/test_cnn.py]: Is this not done already in your fixture? Perhaps you called super too early?

File: tests/garage/tf/core/test_cnn.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    def test_shape(self):
```
Comment (ryanjulian) [inline tests/garage/tf/core/test_cnn.py]: Please assert something about actual input/output pairs rather than just shape.

File: garage/tf/core/parameter.py (hunk 8 above)
Comment (ahtsan) [reply garage/tf/core/parameter.py]: It is from @zhanpenghe

Comment (ryanjulian) [reply garage/tf/core/parameter.py]: You are responsible for every line of code you commit to this repository, regardless of its origin.

File: garage/tf/core/cnn.py (hunk 6 above)
Comment (ahtsan) [reply garage/tf/core/cnn.py]: In many cases people do not use max pooling layers and in that case, pool shape is not needed. What should be a good way for that?

File: tests/garage/tf/core/test_cnn.py (hunk 13 above)
Comment (ahtsan) [reply tests/garage/tf/core/test_cnn.py]: I didn't see this in `TfGraphTestCase`. Should I add this there?

File: tests/garage/tf/core/test_cnn.py (hunk 14 above)
Comment (ahtsan) [reply tests/garage/tf/core/test_cnn.py]: Do you mean to obtain the same output given an input without using tensorflow conv2d?

File: tests/garage/tf/core/test_cnn.py (hunk 13 above)
Comment (ahtsan) [reply tests/garage/tf/core/test_cnn.py]: `TfGraphTestCase` is for testing graphs, where `tf.global_variables_initializer()` cannot be called in `setUp()`. I can create a new fixture for testing trivial things and initialize in `setUp()`.

Comment (CatherineSue) [reply tests/garage/tf/core/test_cnn.py]: Is it possible to delete TfTestCase and only retain TfGraphTestCase? @jonashen 
Looks like creating a new graph and pass it to a session is relatively important when we have multiple tests in a test class, as in Anson's tests. And I can't find other tests that use TfTestCase.

We may as well call `tf.global_variables_initializer()` in setUp. Some algorithms test cases also use TfGraphTestCase. And algo.train() also calls `tf.global_variables_initializer()`. But it looks ok to call it twice in those cases.

Comment (ryanjulian) [reply tests/garage/tf/core/test_cnn.py]: I defer to @CatherineSue 's judgement

File: garage/tf/core/cnn.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+    with tf.variable_scope(name):
+        h = input_var
+        for index, (filter_dim, num_filter) in enumerate(
+                zip(filter_dims, num_filters)):
+            h = hidden_nonlinearity(
+                _conv(h, 'h{}'.format(index), filter_dim, num_filter, strides,
+                      hidden_w_init, hidden_b_init, padding))
+            if max_pooling:
+                h = tf.nn.max_pool(
```
Comment (ryanjulian) [inline garage/tf/core/cnn.py]: you don't test the max_pooling case. please add a test to cover this

File: garage/tf/core/parameter.py (hunk 9 above)
Comment (eric-heiden) [reply garage/tf/core/parameter.py]: Typo ("Paramter")

File: garage/tf/core/parameter.py (hunk 8 above)
Comment (eric-heiden) [reply garage/tf/core/parameter.py]: Typo ("Paramter")

File: garage/tf/core/parameter.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
+"""Parameter function with tensorflow as the only dependency."""
```
Comment (eric-heiden) [inline garage/tf/core/parameter.py]: "TensorFlow"

File: garage/tf/q_functions/discrete_mlp_q_function.py (hunk 11 above)
Comment (eric-heiden) [reply garage/tf/q_functions/discrete_mlp_q_function.py]: Discrelte

File: tests/garage/tf/core/test_cnn.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+        # flatten
+        h_out = np.full(
+            (self.batch_size, 8 * 8 * self.out_channels[-1]),
+            filter_sum,
+            dtype=np.float32)
+        # pass to a dense layer
+        dense_in = tf.matmul(h_out, out_w) + out_b
+        np.testing.assert_array_equal(dense_in.eval(), result)
+
+    def convolve(self, input, filter_weights, filter_biass, stride):
```
Comment (CatherineSue) [inline tests/garage/tf/core/test_cnn.py]: Please use a different name for `input`.

Comment (cpnota) [reply tests/garage/tf/core/test_cnn.py]: Any specific reason? In the generic case `input` is a bad name for a function parameter, but in this case `convolve` is a mathematical function of `input`, and it's helpful to know which parameter is the "input" to the function being represented.

Comment (ryanjulian) [reply tests/garage/tf/core/test_cnn.py]: `input()` is a [Python builtin](https://docs.python.org/3/library/functions.html), so using the variable name `input` masks the symbol in this context.

PEP8 [gently recommends](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles) using a trailing underscore to avoid overloading builtins. I've also used `an_input` in the past.

File: garage/tf/core/networks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryanjulian) [reply garage/tf/core/networks.py]: yeah that sounds fine.

**Comment by ahtsan** [comment]
@ryanjulian can you give a pass on it?

File: tests/garage/tf/core/test_cnn.py [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+        # flatten
+        h_out = np.full(
+            (self.batch_size, 8 * 8 * self.out_channels[-1]),
+            filter_sum,
+            dtype=np.float32)
+        # pass to a dense layer
+        dense_in = tf.matmul(h_out, out_w) + out_b
+        np.testing.assert_array_equal(dense_in.eval(), result)
+
+    def convolve(self, _input, filter_weights, filter_biass, stride):
```
Comment (ryanjulian) [inline tests/garage/tf/core/test_cnn.py]: i believe numpy and scipy provide a convolution function. If you can use that instead, you should
